### PR TITLE
Make the payment intent creation endpoint backend agnostic

### PR DESCRIPTION
**Pull request by gregorbg**
Same idea as #9539 

Partial implementation of #9208

### CODE REVIEW

File: app/models/paypal_record.rb [hunk 1]
```diff
@@ -50,6 +61,11 @@ class PaypalRecord < ApplicationRecord
   belongs_to :parent_record, class_name: "PaypalRecord", inverse_of: :child_records, optional: true
   has_many :child_records, class_name: "PaypalRecord", inverse_of: :parent_record, foreign_key: :parent_record_id
 
+  def determine_wca_status
+    result = WCA_TO_PAYPAL_STATUS_MAP.find { |key, values| values.include?(self.paypal_status) }
+    result&.first || raise("No associated wca_status for paypal_status: #{self.paypal_status} - our tests should prevent this from happening!")
```
Comment (dunkOnIT) [inline app/models/paypal_record.rb:66]: Do we have a test that prevents this from happening? I believe we have a test for Stripe that asserts that all wca_statuses have a corresponding stripe_status, but I don't see once in this PR for paypal (apologies if I'm just being dense)

File: app/views/registrations/_paypal_payment_form.html.erb [hunk 2]
```diff
@@ -103,12 +103,12 @@
...
 
           return wrapAjaxPromise('load-payment-intent', {
-            url: '<%= registration_create_paypal_order_path(@registration) %>',
+            url: '<%= registration_payment_intent_path(@registration, :paypal) %>',
             method: 'POST',
             headers: { 'Content-Type': 'application/json' },
             data: JSON.stringify({ amount: amount }),
           })
-            .then(({ id: orderId }) => orderId)
+            .then(({ client_secret: orderId }) => orderId)
```
Comment (dunkOnIT) [inline app/views/registrations/_paypal_payment_form.html.erb:111]: Why has the name of this variable been changed? I don't see it being invoked anywhere else

**Review by dunkOnIT — COMMENTED** [review]
Generally looks good, just some questions that I want to make sure I understand

File: app/controllers/registrations_controller.rb [hunk 3]
```diff
@@ -3,7 +3,7 @@
 require "csv"
 
 class RegistrationsController < ApplicationController
-  before_action :authenticate_user!, except: [:create, :index, :psych_sheet, :psych_sheet_event, :register, :stripe_webhook, :payment_denomination, :create_paypal_order]
+  before_action :authenticate_user!, except: [:create, :index, :psych_sheet, :psych_sheet_event, :register, :stripe_webhook, :payment_denomination]
```
[the remark this answers is not in the document]
Comment (gregorbg) [reply app/controllers/registrations_controller.rb:6]: It is. You missed the negation hidden in the `except` :upside_down_face:

File: app/views/registrations/_paypal_payment_form.html.erb (hunk 2 above)
Comment (dunkOnIT) [inline app/views/registrations/_paypal_payment_form.html.erb:111]: Change `client_secret` to `render_code` or `payment_render_code`

File: app/models/paypal_record.rb (hunk 1 above)
Comment (gregorbg) [reply app/models/paypal_record.rb:75]: Tests added.

File: app/views/registrations/_paypal_payment_form.html.erb (hunk 2 above)
Comment (gregorbg) [reply app/views/registrations/_paypal_payment_form.html.erb:111]: Create a migration at a later point in time :innocent:

File: spec/models/paypal_record_spec.rb [hunk 4]
```diff
@@ -3,6 +3,29 @@
...
+    it 'contains all wca_statuses' do
+      expect(PaypalRecord::WCA_TO_PAYPAL_STATUS_MAP.keys.sort.map(&:to_s)).to eq(PaymentIntent.wca_statuses.values.sort)
+    end
+
+    it 'contains all paypal_statuses' do
+      expect(PaypalRecord.paypal_statuses.keys.sort).to eq(PaypalRecord::WCA_TO_PAYPAL_STATUS_MAP.values.flatten.sort)
+    end
+  end
+
+  describe 'validates stripe_status' do
```
Comment (dunkOnIT) [inline spec/models/paypal_record_spec.rb:16]: Should be named `validates paypal_status`

Comment (gregorbg) [reply spec/models/paypal_record_spec.rb:16]: Aaaargh copy-pasta :spaghetti:

**Review by dunkOnIT — APPROVED** [review]
:tada:

**Pull request merged**